client/implementation-vertx/src/main/java/io/smallrye/graphql/client/vertx/typesafe/VertxTypesafeGraphQLClientBuilder.java

```diff
@@ -55,39 +58,47 @@ public TypesafeGraphQLClientBuilder endpoint(URI endpoint) {
         return this;
     }
 
+    @Override
+    public TypesafeGraphQLClientBuilder header(String name, String value) {
+        if (this.headers == null) {
+            this.headers = new LinkedHashMap<>();
+        }
+        this.headers.put(name, value);
+        return this;
+    }
+
     @Override
     public <T> T build(Class<T> apiClass) {
         if (configKey == null) {
             configKey = configKey(apiClass);
         }
 
-        GraphQLClientsConfiguration configs = GraphQLClientsConfiguration.getInstance();
-        GraphQLClientConfiguration persistentConfig = configs
-                .getClient(configKey);
-        if (persistentConfig == null) {
-            // in case that we're running in a plain Java SE application, it is possible that the client configuration
-            // hasn't been added yet, because there is no CDI extension or Jandex processor that scans for @GraphQLClientApi annotations
-            // at startup => try adding a configuration entry dynamically for this client interface in particular.
-            // Then try again.
-            configs.addTypesafeClientApis(Collections.singletonList(apiClass));
-            persistentConfig = configs
-                    .getClient(configKey);
-        }
-        if (persistentConfig != null) {
-            applyConfig(persistentConfig);
-        }
+        applyConfigFor(apiClass);
 
         if (endpoint == null) {
             throw ErrorMessageProvider.get().urlMissingErrorForNamedClient(configKey);
         }
 
         initClients();
-        VertxTypesafeGraphQLClientProxy graphQlClient = new VertxTypesafeGraphQLClientProxy(persistentConfig,
-                endpoint, httpClient, webClient);
+        VertxTypesafeGraphQLClientProxy graphQlClient = new VertxTypesafeGraphQLClientProxy(headers, endpoint, httpClient,
+                webClient);
         return apiClass.cast(Proxy.newProxyInstance(getClassLoader(apiClass), new Class<?>[] { apiClass },
                 (proxy, method, args) -> invoke(apiClass, graphQlClient, method, args)));
     }
 
+    private void applyConfigFor(Class<?> apiClass) {
+        GraphQLClientsConfiguration configs = GraphQLClientsConfiguration.getInstance();
+        // In case that we're running in a plain Java SE application, it is possible that the client configuration
+        // hasn't been added yet, because there is no CDI extension or Jandex processor that scans for @GraphQLClientApi annotations
+        // at startup => try adding a configuration entry dynamically for this client interface in particular.
+        configs.initTypesafeClientApi(apiClass);
```

Comment:
Yes, we do this always, so _all_ the config is being picked up, i.e. also credentials

Reply:
If you do this repeatedly (without checking that the config is not initialized yet) you will overwrite the customization that Quarkus performs on the configuration class - using `io.quarkus.smallrye.graphql.client.runtime.GraphQLClientConfigurationMergerBean` which merges config added through  native `quarkus.*` properties into the configuration class, so configuration using `quarkus.*` properties will stop working - this is why you're getting errors in the Quarkus tests.

It seems fixed if I do this:
```
private void applyConfigFor(Class<?> apiClass) {
        GraphQLClientsConfiguration configs = GraphQLClientsConfiguration.getInstance();

        GraphQLClientConfiguration persistentConfig = configs.getClient(configKey);
        if (persistentConfig == null) {
            // In case that we're running in a plain Java SE application, it is possible that the client configuration
            // hasn't been added yet, because there is no CDI extension or Jandex processor that scans for @GraphQLClientApi annotations
            // at startup => try adding a configuration entry dynamically for this client interface in particular.
            configs.initTypesafeClientApi(apiClass);
            persistentConfig = configs.getClient(configKey);
        }
        if (persistentConfig != null) {
            applyConfig(persistentConfig);
        }
    }
```

I'm not sure what you mean by *all* config not getting picked up, do you see any problem with this?
SmallRye should generally call `initTypesafeClientApi` only in a plain Java SE app - in Quarkus, this is called by Quarkus, and if you call it again, you overwrite what Quarkus enhanced.

Reply:
Oh, I think I know what you mean now - the tests in `ConfigBehavior` change the properties-based configuration over time, so you need to rescan the config for the same typesafe client class. That's a bit problematic because we generally assume the configuration to be immutable and not expected to change, at least Quarkus always treats it as such.

Reply:
Would it be too much of a hassle for you to rewrite the tests so that you don't change the configuration for one typesafe client over time?
Especially since it's a TCK, and therefore it's meant to be reused by different implementations in the future, I think we should not assume that the implementation will consider the config mutable, especially in the era of microservice toolkits like Quarkus and Micronaut.

Reply:
I get your point. I don't want to mutate existing config, but merge different config "sources", i.e. annotations and config properties. I'll have to think about how to solve this properly. Thanks for investigating!

Reply:
now it works... WDYT?

Reply:
Ok this will do